### PR TITLE
Inheritance cleanups

### PR DESCRIPTION
**Pull request by edwardalee**
This PR changes the way inheritance is handled to fix two problems identified in #907.

### CODE REVIEW

**Review by lhstrh — APPROVED** [review]
Looks good! I believe there are some unit tests that @housengw commented out because of the buggy behavior that is now fixed. If my recollection is correct, we should re-enable those unit tests.

**Comment by lhstrh** [comment]
Looks like my changes introduced a bug. Sorry!

File: org.lflang/src/org/lflang/ASTUtils.java [hunk 1]
```diff
@@ -611,6 +542,29 @@ public static String getUniqueIdentifier(Reactor reactor, String name) {
...
+     * a given reactor definition.
+     * @param definition The reactor definition.
+     * @param elements A function that maps a reactor definition to a list of
+     *                 elements of type T.
+     * @param <T> The type of elements to collect (e.g., Port, Timer, etc.)
+     * @return
+     */
+    public static <T> List<T> collectElements(Reactor definition, Function<Reactor,List<T>> elements) {
+        List<T> result = new ArrayList<T>();
+        // Add elements of the current reactor.
```
Comment (cmnrd) [inline org.lflang/src/org/lflang/ASTUtils.java:556]: This change looks great! The only "mistake" that I could spot is that the elements of the current reactor are added before for the elements of the super classes. In the original code, it was added after the super class elements. I don't know whether the order within the list has any meaning in this context. If it should not have meaning, then maybe the bug lies somewhere else.

**Review by cmnrd — COMMENTED** [review]
The changes look good to me.

Indeed, the reordering in `collectElements` fixed the tests. As stated in my comment above, the fact that the test outcome depends on the order of elements within the resulting list, might indicate a bug further down the line.

**Comment by edwardalee** [comment]
The order very definitely has meaning.  Not only should the base class elements be added first, but also when a reactor extends multiple base classes, those elements need to be added in the order that the extensions are declared. So there is only exactly one order that is correct.

Consider a reaction that writes to an output.  A subclass can "override" the output of a base class by overwriting it.

I'm surprised my test caught this bug. I didn't design it to check for this...

**Pull request merged**